src/compiler/checker.ts

```diff
@@ -14748,21 +14748,70 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+    function getSignaturesOfType(typeIn: Type, kind: SignatureKind): readonly Signature[] {
+        const reducedType = getReducedApparentType(typeIn);
+        function carveoutResult(): readonly Signature[] | undefined {
+            if (kind === SignatureKind.Call && reducedType.flags & TypeFlags.Union) {
+                // If the union is all different instantiations of a member of the global array type...
+                let memberName: __String;
+                if (everyType(reducedType, t => !!t.symbol?.parent && isArrayOrTupleSymbol(t.symbol.parent) && (!memberName ? (memberName = t.symbol.escapedName, true) : memberName === t.symbol.escapedName))) {
+                    if ((reducedType as UnionType).arrayFallbackSignatures) {
+                        return (reducedType as UnionType).arrayFallbackSignatures!;
+                    }
+
+                    // calculate return types as union of return types over the associated type, rather than return type associated with the union of types
+                    const numTypes = (reducedType as UnionType).types.length;
+                    const numSigs = (reducedType as UnionType).types[0].symbol.declarations?.length;
+                    Debug.assert(numSigs);
+                    const returnTypes = (new Array(numSigs).fill(/*value*/ undefined)).map((_, isig): Type | undefined => {
+                        // getUnionType(*,UnionReduction.Subtype) doesn't detect duplicate types (unless I'm mistaken) so use set first.
+                        const rtset = new Set<Type>();
+                        new Array(numTypes).fill(/*value*/ undefined).forEach((_, itype) => {
+                            // Notice we potentially generate the whole signature to get the return type.
+                            // Maybe that could be avoided?
+                            const callSignatures = ((reducedType as UnionType).types[itype] as ObjectType).callSignatures
+                                ?? getSignaturesOfStructuredType((reducedType as UnionType).types[itype] as ObjectType, SignatureKind.Call);
+                            if (callSignatures[isig].typeParameters) return; // skip generic signatures, i.e. <S>
+                            const returnType = callSignatures[isig].resolvedReturnType
+                                ?? getReturnTypeOfSignature(callSignatures[isig]);
+                            Debug.assert(returnType);
+                            rtset.add(returnType);
+                        });
+                        const art: Type[] = [];
+                        rtset.forEach(rt => art.push(rt));
+                        if (art.length === 0) return undefined;
+                        if (art.length === 1) return art[0];
+                        return getUnionType(art, UnionReduction.Subtype);
+                    });
+
+                    // Transform the type from `(A[] | B[])["member"]` to `(A | B)[]["member"]` (since we pretend array is covariant anyway)

```

Comment:
The original carveout code description - "// Transform the type from `(A[] | B[])["member"]` to `(A | B)[]["member"]` "
In this pull that action is kept for the signature parameters, but the return type is reverted the union of return types, where appropriate.